Allow verification of a specified backup.

silent-observer opened the pull request:
This patch adds support for `--set` option for the `verify` command. This option was already accepted without errors, but wasn't actually implemented (see #2419).

Default behavior for `verify` is to check all the backups present. With new `--set` option only the specified backup will be verified (similar to how `--set` works for `restore`). If the specified backup label is not valid an error is added to the result and verification is skipped.

dwsteele reviewed (changes requested):
I tried this out and it worked fine for backups without block incremental but it failed when I enabled block incremental. First I setup an block incremental with changes:
```
$ pgbackrest/doc/doc.pl --out=html --include=user-guide --require=/quickstart/check-$ configuration --pre --no-cache --var=encrypt=n --var=host-mem-limit=n
$ docker exec -it -u postgres doc-pg-primary pgbackrest --stanza=demo backup --$ start-fast --type=full --repo1-block --repo1-bundle --repo1-bundle-limit=8k --repo1-retention-full=1
$ docker exec -it -u postgres doc-pg-primary psql -t -c 'create table test (id int)'
$ docker exec -it -u postgres doc-pg-primary pgbackrest --stanza=demo backup --start-fast --type=diff --repo1-block --repo1-bundle --repo1-bundle-limit=8k
```
Then I deleted a required file from the full but verify still succeeded:
```
$ docker exec -it -u postgres doc-pg-primary rm /var/lib/pgbackrest/backup/demo/20250217-213926F/pg_data/base/5/2704.pgbi
$ docker exec -it -u postgres doc-pg-primary pgbackrest --stanza=demo verify --log-level-console=info --set=20250217-213926F_20250217-213934D
P00   INFO: verify command begin 2.55.0dev: --exec-id=464-1f04d93f --log-level-console=info --no-log-timestamp --repo1-path=/var/lib/pgbackrest --set=20250217-213926F_20250217-213934D --stanza=demo
P00   INFO: verify command end: completed successfully
```
Then I did a restore of the same backup and it failed because of the missing file:
```
$ docker exec -it -u postgres doc-pg-primary pg_ctlcluster 15 demo stop
$ docker exec -it -u postgres doc-pg-primary rm -rf /var/lib/postgresql/15/demo
$ docker exec -it -u postgres doc-pg-primary pgbackrest restore --stanza=demo --set=20250217-213926F_20250217-213934D
P00  ERROR: [055]: raised from local-1 protocol: unable to open missing file '/var/lib/pgbackrest/backup/demo/20250217-213926F/pg_data/base/5/2704.pgbi' for read
                [FileMissingError] on 2 retries from 5-15012ms: unable to open missing file '/var/lib/pgbackrest/backup/demo/20250217-213926F/pg_data/base/5/2704.pgbi' for read
```
The verify code is fooled because there is no reference in the incremental for 2704.pgbi so it thinks there are no dependencies, which would have been true before block incremental. But with block incremental the references are stored in the block map.

The easiest thing to do here would be to build a list of all backups that the backup specified by `--set` could depend on and check them all. Not ideal of course, but it would work in all cases. I'd probably recommend this for a first pass since it is still better than what we have.

The more complicated way would be to read all the maps and check just the blocks that would be required to reconstruct the specified backup -- essentially what restore does. Hopefully we could reuse some of the restore code for that.

dwsteele reviewed (changes requested):
There seem to be a few problems here. From reading the code I could see that it would not work correctly if block incremental is disabled. After the diff backup I ran:
```
docker exec -it -u postgres doc-pg-primary pgbackrest --stanza=demo backup --start-fast --type=incr
docker exec -it -u postgres doc-pg-primary pgbackrest --stanza=demo verify --log-level-console=info --set=20250226-221003F_20250226-221406I
```
I expect it to succeed even with the missing file because block-incr had been disabled and the verify code would not know it was using for some files. However, I got this instead:
```
P00   INFO: verify command begin 2.55.0dev: --exec-id=535-422b0676 --log-level-console=info --no-log-timestamp --repo1-path=/var/lib/pgbackrest --set=20250226-221003F_20250226-221406I --stanza=demo
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/13342.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/13347.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/13352.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/13357.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/1417.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/1418.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2224.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2328.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2336.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2604.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2611.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2613.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2620.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2830.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2832.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2834.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/2995.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3118.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3256.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3350.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3381.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3429.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3430.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3439.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3466.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3501.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3576.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3596.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/3598.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4143.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4145.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4147.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4149.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4151.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4153.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4155.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4157.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4159.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4163.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4165.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4167.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4169.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4171.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/4173.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6102.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6104.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6106.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6175.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6228.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/6237.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/1/826.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/13342.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/13347.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/13352.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/13357.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/1417.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/1418.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2224.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2328.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2336.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2604.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2611.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2613.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2620.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2830.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2832.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2834.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/2995.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3118.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3256.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3350.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3381.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3429.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3430.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3439.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3466.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3501.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3576.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3596.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/3598.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4143.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4145.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4147.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4149.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4151.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4153.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4155.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4157.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4159.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4163.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4165.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4167.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4169.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4171.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/4173.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6102.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6104.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6106.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6175.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6228.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/6237.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/4/826.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/13342.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/13347.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/13352.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/13357.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/1417.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/1418.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/16385.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2224.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2328.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2336.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2604.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2611.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2613.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2620.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2830.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2832.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2834.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/2995.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3118.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3256.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3350.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3381.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3429.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3430.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3439.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3466.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3501.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3576.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3596.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/3598.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4143.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4145.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4147.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4149.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4151.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4153.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4155.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4157.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4159.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4163.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4165.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4167.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4169.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4171.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/4173.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6102.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6104.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6106.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6175.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6228.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/6237.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/base/5/826.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/1214.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/2846.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/2964.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/2966.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/3592.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4060.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4175.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4177.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4181.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4183.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/4185.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/6000.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/6100.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/6243.gz'
P01   INFO: file missing '20250226-221003F_20250226-221318I/pg_data/global/6244.gz'
P00   INFO: stanza: demo
            status: error
              backup: 20250226-221003F_20250226-221406I, status: invalid, total files checked: 962, total valid files: 793
                missing: 169
P00   INFO: verify command end: completed successfully
```
Definitely something off going on there and I notice that `pg_data/base/5/2704.pgbi` is not in the list so I think that is being missed as well but hard to be sure with the other errors.

In any case, checking the blockIncr flag is not enough since it might be disabled. You'll need to iterate the file list and look for any files that have blockIncr enbaled.

silent-observer commented:
I think I fixed it. The errors were caused by incorrect ordering of backups in the list, so I added sorting back.

As for checking for blockIncr, did you mean like this, checking `blockIncrMapSize` for every file?

dwsteele commented:
> As for checking for blockIncr, did you mean like this, checking blockIncrMapSize for every file?

I should have stated it more clearly, but yeah, that's what I meant.

dwsteele reviewed (changes requested):
The issues so far have been addressed. Thanks! I've done some review and cleanup in https://github.com/dwsteele/pgbackrest/tree/ADBM-1978-cig so please pull those changes in.

A new thing I've noticed is that all WAL in all archive IDs is being checked when --set is specified. For example, if I make a full and diff backup then corrupt a WAL segment only needed for the full backup I'll still get an error when verifying the diff backup, for example:
```
$ pgbackrest --stanza=demo verify --log-level-console=info --set=20250301-220154F_20250301-220202D
P00   INFO: verify command begin 2.55.0dev: --exec-id=593-496e1b32 --log-level-console=info --no-log-timestamp --repo1-path=/var/lib/pgbackrest --set=20250301-220154F_20250301-220202D --stanza=demo
P01   INFO: invalid checksum '15-1/0000000100000000/00000001000000000000000B-87971e85b443d0e1b53fc3c5975b734a825705df.gz'
P00   INFO: stanza: demo
            status: error
              archiveId: 15-1, total WAL checked: 3, total valid WAL: 2
                checksum invalid: 1
P00   INFO: verify command end: completed successfully
```
I think we should limit WALs checked to the archive ID related to the backup and *at least* start our WAL verification with the first WAL in the selected backup. Then the question is, do we check WAL all the way to the end or stop at the last WAL for the selected backup?

Verifying all the way to the end would be safer but could take a lot longer than the user expects so maybe it makes sense to just check the WAL range required to make the backup consistent. Thoughts?

silent-observer commented:
> The issues so far have been addressed. Thanks! I've done some review and cleanup in https://github.com/dwsteele/pgbackrest/tree/ADBM-1978-cig so please pull those changes in.

Thank you!

> I think we should limit WALs checked to the archive ID related to the backup and at least start our WAL verification with the first WAL in the selected backup. Then the question is, do we check WAL all the way to the end or stop at the last WAL for the selected backup?

I've actually been working on this already, I think checking only WALs related to the backup would be better. --set is already supposed to improve verification time, so better performance here is more important, I think. Also, on a particularly busy system there is a possibility that WALs are being added faster than they are verified, so it would be useful to verify only the specified WAL range. I'll provide my current WAL filtering solution here.

On the side note, am I correct that only the main backup's WAL range has to be checked, not WAL ranges of its dependencies?

dwsteele commented:
> I think checking only WALs related to the backup would be better. --set is already supposed to improve verification time, so better performance here is more important, I think

That's pretty much my thinking. Let's just check the backup range, then.

> On the side note, am I correct that only the main backup's WAL range has to be checked, not WAL ranges of its dependencies?

You are correct. Only WAL for the specified backup is needed to recover that backup.

dwsteele commented:
It will be a few weeks before I can look at this again but I'm not ignoring it.

dwsteele reviewed (approved):
I made some changes to avoid redundant loads of backup.manifest and to avoid output of the same errors multiple times. Please pull in the changes from https://github.com/dwsteele/pgbackrest/tree/ADBM-1978-cig.

I'm not too happy with how WAL verification is being handled but that is not the fault of this commit so I'm OK with proceeding to commit. Thanks for the patch!

silent-observer commented:
Thank you!
